### ISSUE TITLE
insert failed

### ISSUE THREAD

**Issue by ruiluis** [akumuli/Akumuli#305]
i run the tests in functest directory using normal and bulk methods worked ok.
then i try run with my data model, that is composed of 4 measurements and can have around 60 tags and it keeps giving core dump. the message size is around 700 or more characters..
There is a limit on the number of possible tags?
How can i run in Debug mode?

**Comment by Lazin** [comment]
Could you send me the input sample that triggers core dump?
As far as I recall there is no limit on number of tags but there is a limit on series name length. It can't exceed 4KB.

**Comment by ruiluis** [comment]
it does not happen in the same entry.. it happens after inserting some entries, but is random.. i even try to open and close the connection several times when running but with no improvement.

it crashed when adding this entry:
+aff_revenue|revenue|referral_commission|sale_sum_tti country_code=119 idaccmanager=20189 idaffiliate=1452131052 idaffmanager=20176 idapplication=1075048 idbrowser=7 idcampaign=1075048 idhardware=2 idoperator=0 mnc=0 mcc=0 idos=1 idpartner=5093 idurl=0 ip=58.78.128.30 is_proxy=0 vhost=test.com account_type=21 browser_version=73.0.3683.90 model=LM-V409N vendor=LG net16=58.78 net24=58.78.128 afc_website_id=102 afc_placement=122674731522 browser_name=Chrome\ Mobile os_name=Android os_version=8.1.0 country_iso3166=KR idcategory=34 idtenant=4 city=ulsan tzoffs=+900 offer_type=1 eye_score=1 eye_sale_score=1
+20190330T044208.000000
*4
+1.176252
+1.96042
+0
+427
it give a core dump when adding this entry

**Comment by Lazin** [comment]
The message looks good, I'll try to reproduce and let you know. BTW, can you share log output?

**Comment by ruiluis** [comment]
2019-05-14 16:13:44,537 [0x7f46e2cc1700] Main [TRACE] INSERT INTO akumuli_series (series_id, keyslist, storage_id)
	SELECT 'revenue' as series_id, 'account_type=17 afc_placement=10951_59954_44551_32230 afc_website_id=95 browser_name=Safari browser_version=11.0 city=bensheim country_code=56 country_iso3166=DE idaccmanager=1938 idaffiliate=1452146859 idaffmanager=20113 idapplication=1073989 idbrowser=32 idcampaign=1073989 idcategory=73 idhardware=2 idoperator=0 idos=2 idpartner=5128 idtenant=4 idurl=0 ip=217.113.186.59 is_proxy=0 mcc=0 mnc=0 model=iPhone net16=217.113 net24=217.113.186 os_name=iOS os_version=12_0 vendor=Apple vhost=test.com' as keyslist, 11752  as storage_id
	UNION SELECT 'referral_commission', 'account_type=17 afc_placement=10951_59954_44551_32230 afc_website_id=95 browser_name=Safari browser_version=11.0 city=bensheim country_code=56 country_iso3166=DE idaccmanager=1938 idaffiliate=1452146859 idaffmanager=20113 idapplication=1073989 idbrowser=32 idcampaign=1073989 idcategory=73 idhardware=2 idoperator=0 idos=2 idpartner=5128 idtenant=4 idurl=0 ip=217.113.186.59 is_proxy=0 mcc=0 mnc=0 model=iPhone net16=217.113 net24=217.113.186 os_name=iOS os_version=12_0 vendor=Apple vhost=test.com', 11753
	UNION SELECT 'sale_sum_tti', 'account_type=17 afc_placement=10951_59954_44551_32230 afc_website_id=95 browser_name=Safari browser_version=11.0 city=bensheim country_code=56 country_iso3166=DE idaccmanager=1938 idaffiliate=1452146859 idaffmanager=20113 idapplication=1073989 idbrowser=32 idcampaign=1073989 idcategory=73 idhardware=2 idoperator=0 idos=2 idpartner=5128 idtenant=4 idurl=0 ip=217.113.186.59 is_proxy=0 mcc=0 mnc=0 model=iPhone net16=217.113 net24=217.113.186 os_name=iOS os_version=12_0 vendor=Apple vhost=test.com', 11754
	UNION SELECT 'aff_revenue', 'account_type=3 afc_website_id=59e1ae469bdb9a3589d77b05 browser_lang=ko browser_name=Chrome\ Mobile browser_version=61.0.3163.98 city=wonju country_code=119 country_iso3166=KR idaccmanager=20189 idaffiliate=1452127559 idaffmanager=20176 idapplication=1075187 idbrowser=7 idcampaign=1075187 idcategory=34 idhardware=2 idoperator=0 idos=1 idpartner=5095 idtenant=4 idurl=0 ip=121.187.9.91 is_proxy=0 mcc=0 mnc=0 model=SM-G935K net16=121.187 net24=121.187.9 os_name=Android os_version=8.0.0 vendor=Samsung vhost=test.com', 11755
;
2019-05-14 16:13:44,537 [0x7f46e2cc1700] Main [TRACE] INSERT OR REPLACE INTO akumuli_rescue_points (storage_id, addr0, addr1, addr2, addr3, addr4, addr5, addr6, addr7) VALUES ( 11755, null, null, null, null, null, null, null, null),( 11754, null, null, null, null, null, null, null, null),( 11752, null, null, null, null, null, null, null, null),( 11753, null, null, null, null, null, null, null, null);
2019-05-14 16:13:44,537 [0x7f46e2cc1700] Main [TRACE] END TRANSACTION;
2019-05-14 16:13:44,538 [0x7f46e2cc1700] Main [TRACE] BEGIN TRANSACTION;
2019-05-14 16:13:44,538 [0x7f46e2cc1700] Main [TRACE] INSERT INTO akumuli_series (series_id, keyslist, storage_id)
	SELECT 'revenue' as series_id, 'account_type=3 afc_website_id=59e1ae469bdb9a3589d77b05 browser_lang=ko browser_name=Chrome\ Mobile browser_version=61.0.3163.98 city=wonju country_code=119 country_iso3166=KR idaccmanager=20189 idaffiliate=1452127559 idaffmanager=20176 idapplication=1075187 idbrowser=7 idcampaign=1075187 idcategory=34 idhardware=2 idoperator=0 idos=1 idpartner=5095 idtenant=4 idurl=0 ip=121.187.9.91 is_proxy=0 mcc=0 mnc=0 model=SM-G935K net16=121.187 net24=121.187.9 os_name=Android os_version=8.0.0 vendor=Samsung vhost=test.com' as keyslist, 11756  as storage_id
	UNION SELECT 'referral_commission', 'account_type=3 afc_website_id=59e1ae469bdb9a3589d77b05 browser_lang=ko browser_name=Chrome\ Mobile browser_version=61.0.3163.98 city=wonju country_code=119 country_iso3166=KR idaccmanager=20189 idaffiliate=1452127559 idaffmanager=20176 idapplication=1075187 idbrowser=7 idcampaign=1075187 idcategory=34 idhardware=2 idoperator=0 idos=1 idpartner=5095 idtenant=4 idurl=0 ip=121.187.9.91 is_proxy=0 mcc=0 mnc=0 model=SM-G935K net16=121.187 net24=121.187.9 os_name=Android os_version=8.0.0 vendor=Samsung vhost=test.com', 11757
	UNION SELECT 'sale_sum_tti', 'account_type=3 afc_website_id=59e1ae469bdb9a3589d77b05 browser_lang=ko browser_name=Chrome\ Mobile browser_version=61.0.3163.98 city=wonju country_code=119 country_iso3166=KR idaccmanager=20189 idaffiliate=1452127559 idaffmanager=20176 idapplication=1075187 idbrowser=7 idcampaign=1075187 idcategory=34 idhardware=2 idoperator=0 idos=1 idpartner=5095 idtenant=4 idurl=0 ip=121.187.9.91 is_proxy=0 mcc=0 mnc=0 model=SM-G935K net16=121.187 net24=121.187.9 os_name=Android os_version=8.0.0 vendor=Samsung vhost=test.com', 11758
	UNION SELECT 'aff_revenue', 'account_type=3 afc_placement=adspub35662 afc_website_id=48 browser_lang=en browser_name=Chrome\ Mobile browser_version=71.0.3578 country_code=225 country_iso3166=US idaccmanager=1938 idaffiliate=1452123426 idaffmanager=21515 idapplication=1073975 idbrowser=7 idcampaign=1073975 idcategory=189 idhardware=2 idoperator=0 idos=1 idpartner=5059 idtenant=4 idurl=0 ip=108.184.118.133 is_proxy=0 mcc=0 mnc=0 model=SM-G570Y net16=108.184 net24=108.184.118 os_name=Android os_version=8.0 vendor=Samsung vhost=test.com', 11759
;
2019-05-14 16:13:44,538 [0x7f46e2cc1700] Main [TRACE] INSERT OR REPLACE INTO akumuli_rescue_points (storage_id, addr0, addr1, addr2, addr3, addr4, addr5, addr6, addr7) VALUES ( 11759, null, null, null, null, null, null, null, null),( 11758, null, null, null, null, null, null, null, null),( 11756, null, null, null, null, null, null, null, null),( 11757, null, null, null, null, null, null, null, null);
2019-05-14 16:13:44,538 [0x7f46e2cc1700] Main [TRACE] END TRANSACTION;
2019-05-14 16:13:44,538 [0x7f46e2cc1700] Main [TRACE] BEGIN TRANSACTION;
2019-05-14 16:13:44,538 [0x7f46e2cc1700] Main [ERROR] Error executing query
2019-05-14 16:13:44,538 [0x7f46e2cc1700] Main [ERROR] near "country_code": syntax error
2019-05-14 16:13:44,538 [0x7f46e2cc1700] Main [ERROR] Terminating (core dumped)

there is something very strange as:
UNION SELECT 'sale_sum_tti', 'account_type=3 afc_website_id=59e1ae469bdb9a3589d77b05 browser_lang=ko browser_name=Chrome\ Mobile browser_version=61.0.3 ......
UNION SELECT 'aff_revenue', 'account_type=3 afc_placement=adspub35662 afc_website_id=48
i have replaced all empty spaces by '\ ' can this be cause by some other character?

**Comment by ruiluis** [comment]
i deleted and create the db.
added the following entries:
+aff_revenue|revenue|referral_commission|sale_sum_tti country_code=222 idaccmanager=1938 idaffiliate=1452123426 idaffmanager=21515 idapplication=1073785 idbrowser=3 idcampaign=1073785 idhardware=2 idoperator=0 mnc=0 mcc=0 idos=1 idpartner=5033 idurl=0 ip=176.36.144.51 is_proxy=0 vhost=test.com account_type=3 model=Redmi\ 5\ Plus vendor=Xiaomi net16=176.36 net24=176.36.144 afc_website_id=9 afc_placement=20310296 browser_name=Android\ Browser os_name=Android os_version=8.1.0 country_iso3166=UA idcategory=33 idtenant=4 city=kiev tzoffs=+200 offer_type=2 eye_score=1 eye_sale_score=1
+20190327T102450.000000
*4
+0
+0
+0
+73

+aff_revenue|revenue|referral_commission|sale_sum_tti country_code=56 idaccmanager=1938 idaffiliate=1452147031 idaffmanager=20164 idapplication=1074317 idbrowser=32 idcampaign=1074317 idhardware=2 idoperator=0 mnc=0 mcc=0 idos=2 idpartner=5149 idurl=0 ip=89.1.215.127 is_proxy=0 vhost=test.com account_type=0 browser_version=12.0 model=iPhone vendor=Apple refdomain=adcan.win net16=89.1 net24=89.1.215 afc_website_id=563621 browser_name=Safari browser_lang=en os_name=iOS os_version=12_0 country_iso3166=DE idcategory=38 idtenant=4 city=hurth tzoffs=+100 offer_type=1 eye_score=1 eye_sale_score=5 eye_sale_score_reason=\ -\ Capping\ reason\ 18
+20190327T102739.000000
*4
+0.3014746
+0.4965464
+0
+117
and this is the resulting entries in the log:
019-05-14 16:37:49,619 [0x7fe8e057f700] Main [TRACE] INSERT INTO akumuli_series (series_id, keyslist, storage_id)
	SELECT 'aff_revenue' as series_id, 'account_type=3 afc_placement=20310296 afc_website_id=9 browser_name=Android\ Browser city=kiev country_code=222 country_iso3166=UA idaccmanager=1938 idaffiliate=1452123426 idaffmanager=21515 idapplication=1073785 idbrowser=3 idcampaign=1073785 idcategory=33 idhardware=2 idoperator=0 idos=1 idpartner=5033 idtenant=4 idurl=0 ip=176.36.144.51 is_proxy=0 mcc=0 mnc=0 model=Redmi\ 5\ Plus net16=176.36 net24=176.36.144 os_name=Android os_version=8.1.0 tzoffs=+200 vendor=Xiaomi vhost=test.com' as keyslist, 1024  as storage_id
;
2019-05-14 16:37:49,619 [0x7fe8e057f700] Main [TRACE] INSERT OR REPLACE INTO akumuli_rescue_points (storage_id, addr0, addr1, addr2, addr3, addr4, addr5, addr6, addr7) VALUES ( 1024, null, null, null, null, null, null, null, null);
2019-05-14 16:37:49,619 [0x7fe8e057f700] Main [TRACE] END TRANSACTION;
2019-05-14 16:37:59,628 [0x7fe8e057f700] Main [TRACE] BEGIN TRANSACTION;
2019-05-14 16:37:59,628 [0x7fe8e057f700] Main [TRACE] INSERT INTO akumuli_series (series_id, keyslist, storage_id)
	SELECT 'revenue' as series_id, 'account_type=3 afc_placement=20310296 afc_website_id=9 browser_name=Android\ Browser city=kiev country_code=222 country_iso3166=UA idaccmanager=1938 idaffiliate=1452123426 idaffmanager=21515 idapplication=1073785 idbrowser=3 idcampaign=1073785 idcategory=33 idhardware=2 idoperator=0 idos=1 idpartner=5033 idtenant=4 idurl=0 ip=176.36.144.51 is_proxy=0 mcc=0 mnc=0 model=Redmi\ 5\ Plus net16=176.36 net24=176.36.144 os_name=Android os_version=8.1.0 tzoffs=+200 vendor=Xiaomi vhost=test.com' as keyslist, 1025  as storage_id
	UNION SELECT 'referral_commission', 'account_type=3 afc_placement=20310296 afc_website_id=9 browser_name=Android\ Browser city=kiev country_code=222 country_iso3166=UA idaccmanager=1938 idaffiliate=1452123426 idaffmanager=21515 idapplication=1073785 idbrowser=3 idcampaign=1073785 idcategory=33 idhardware=2 idoperator=0 idos=1 idpartner=5033 idtenant=4 idurl=0 ip=176.36.144.51 is_proxy=0 mcc=0 mnc=0 model=Redmi\ 5\ Plus net16=176.36 net24=176.36.144 os_name=Android os_version=8.1.0 tzoffs=+200 vendor=Xiaomi vhost=test.com', 1026
	UNION SELECT 'sale_sum_tti', 'account_type=3 afc_placement=20310296 afc_website_id=9 browser_name=Android\ Browser city=kiev country_code=222 country_iso3166=UA idaccmanager=1938 idaffiliate=1452123426 idaffmanager=21515 idapplication=1073785 idbrowser=3 idcampaign=1073785 idcategory=33 idhardware=2 idoperator=0 idos=1 idpartner=5033 idtenant=4 idurl=0 ip=176.36.144.51 is_proxy=0 mcc=0 mnc=0 model=Redmi\ 5\ Plus net16=176.36 net24=176.36.144 os_name=Android os_version=8.1.0 tzoffs=+200 vendor=Xiaomi vhost=test.com', 1027
	UNION SELECT 'aff_revenue', 'account_type=0 afc_website_id=563621 browser_lang=en browser_name=Safari browser_version=12.0 country_code=56 country_iso3166=DE idaccmanager=1938 idaffiliate=1452147031 idaffmanager=20164 idapplication=1074317 idbrowser=32 idcampaign=1074317 idcategory=38 idhardware=2 idoperator=0 idos=2 idpartner=5149 idtenant=4 idurl=0 ip=89.1.215.127 is_proxy=0 mcc=0 mnc=0 model=iPhone net16=89.1 net24=89.1.215 os_name=iOS os_version=12_0 refdomain=adcan.win vendor=Apple vhost=test.com', 1028
;
2019-05-14 16:37:59,628 [0x7fe8e057f700] Main [TRACE] INSERT OR REPLACE INTO akumuli_rescue_points (storage_id, addr0, addr1, addr2, addr3, addr4, addr5, addr6, addr7) VALUES ( 1028, null, null, null, null, null, null, null, null),( 1027, null, null, null, null, null, null, null, null),( 1025, null, null, null, null, null, null, null, null),( 1026, null, null, null, null, null, null, null, null);
2019-05-14 16:37:59,628 [0x7fe8e057f700] Main [TRACE] END TRANSACTION;

**Comment by Lazin** [comment]
You escaped spaces correctly and messages are OK as well as the log output.
I tried to reproduce but so far I haven't seen a core dump. Could you tell me what version/commit-hash you had used? Was WAL enabled in configuration? You could also save messages to text file and send me so I could replay them and reproduce the error if it doesn't have any sensitive information.

**Comment by ruiluis** [comment]
i cloned the repository, compiled. i used wal.
i run extra test and the issue is with the tag values being strings, i converted each tag possible value to int and it inserted correctly.
unfortunately this reveal that i cannot use this db due to the nature of my data.. i have a huge cardinality in the tags possible values, my 35k entries had 85k tag values.. and the memory increase by a lot, and even fail to recover sometimes..  
but i also notice that its very good if you had a lower cardinality in the tag values, and i thinks that its a point you should refer when presenting the db
this is a awesome work.. unfortunately i cannot use it

**Comment by Lazin** [comment]
Probably, it's a known issue that was fixed in 0.8.60. Can you confirm this?
The memory requirements is a bit hard to project. First of all, total cardinality can be larger than number of tags (85k is not big but you can get a lot of tag combinations out of these). Every unique time-series needs from 1KB to 16KB of RAM (the latter one is for really huge series but anyway). Also, system cache will use some memory depending on your workload. 
In short, it's feasible to have millions of unique series on commodity hardware with enough RAM. For tens of millions you will need a server with a lot of RAM. I'll add a documentation page with this information.
